Add CMake build support

FtZPetruska opened the pull request:
Thie PR adds support for building with CMake (See #18). It is based on the existing scons script for files and flags used.

The CMake script includes:
- Generation of CMake package files so the library can be used in other CMake projects by using `find_package(libgambatte)`
- Generation of a pkg-config file, so the library can be used in non-CMake projects by using `pkg-config --cflags --libs libgambatte`
- Building both the static and shared libraries at the same time.
- Support for the `testrunner`.

There were a few code changes, mainly aiming at making building and installing more convenient:
- Move `newstate.h` to the public headers, since `gambatte.h` depends on it, it should be public too
- Instead of passing `-DHAVE_X`, prefer generating a `config.h`. This is necessary as `gbint.h` needs to know which of `cstdint` or `stdint.h` are available, but that information would be lost for a consumer of the library.
- Edit the scons script to generate its own `config.h`, so the shell scripts keep working for the time being

The CI script has been completely reworked:
- Run on Ubuntu, macOS, Msys2 (ucrt-x86_64, i686), Windows (clang-cl, clang)
- Run the test runner
- Test the CMake package (from an install and from the build directory)
- Test the pkg-config file

Lastly, the README.md and INSTALL.md have been updated to remove the references to Qt, use CMake, and cover more common Linux distributions. Fixes #5 

---

A few additional notes:
- the macOS CI environment struggles to get the correct libpng (even specifying its prefix manually), so instead `PNG_LIBRARY` and `PNG_PNG_INCLUDE_DIR` are set manually to help the Find module. Causes the tests to SEGFAULT otherwise.
- MSVC currently does NOT work, there is a VLA in `src/statesaver.cpp:553` that breaks compilation. It is probably not a difficult fix, but this PR is already big enough as-is, and there may be more issues.

FtZPetruska committed:
Make newstate a public header.

Since it is included in gambatte.h it needs to be installed.

FtZPetruska committed:
Add common CMake build directories to gitignore.

FtZPetruska committed:
Add CMake build script.

The current implementation provides:
- Options to build both static and shared library at the same time
- A toggle for enabling warnings as errors
- Support for enabling zlib
- Generating pkg-config and CMake targets
- Exporting CMake targets to consume from the build directory

FtZPetruska committed:
Replace defines with a generated config.h.

This is required as public header rely on knowing the value of
HAVE_STDINT_H and HAVE_CSTDINT.

This was also backported to scons since it edits code.

FtZPetruska committed:
CMake: Add support for tests.

FtZPetruska committed:
Update build docs to use CMake.

FtZPetruska committed:
CI: Update workflow to use CMake.

Changes include:
- building on multiple platforms
- run ctest tests
- test consuming from the build directory, from an installation, and
from pkg-config.

CasualPokePlayer commented on INSTALL.md:
```diff
@@ -1,120 +1,139 @@
+
+As there is no standard package manager for Visual Studio, you will need to either manually build zlib (and libpng for test), or use a package manager such as [vcpkg](https://github.com/microsoft/vcpkg).
+
+You will also need to manually install [CMake](https://cmake.org/) and to make sure to add it to your `PATH` during installation.
```
CMake is one of the possible components which can be installed with the Visual Studio installer (C++ CMake tools for Windows) which also includes Ninja.

FtZPetruska replied:
Thank you for the feedback, I've updated the documentation appropriately. It's a similar situation to Msys2 shell where the tools are only available from a specific shell (Developer PowerShell for VS).

CasualPokePlayer commented on INSTALL.md:
```diff
@@ -1,120 +1,139 @@
 
+### Visual Studio users
+
+As there is no standard package manager for Visual Studio, you will need to either manually build zlib (and libpng for test), or use a package manager such as [vcpkg](https://github.com/microsoft/vcpkg).
```
For this, we may want to have a folder containing submodules to the libraries (and maybe add in cmake scripts for them, if they don't already have them). That way they can always be statically included (we just include the cmake script for them).

FtZPetruska replied:
I am not a fan of vendoring for a library, since you cannot embed a static library in another static library, we would need to install them too. Moreover, since Python is technically a dependency to run tests, it's rather impractical to vendor that.

Thankfully, it seems that starting Visual Studio 17.6, installing vcpkg is just an option like CMake. It only supports manifest mode (hence the addition of a `vcpkg.json`) but makes it much more trivial for Visual Studio users to get it setup.

CasualPokePlayer commented on CMakeLists.txt:
```diff
@@ -0,0 +1,308 @@
+  "libgambatte/src/mem/mbc/mmm01.cpp"
+  "libgambatte/src/mem/mbc/pocket_camera.cpp"
+  "libgambatte/src/mem/mbc/wisdom_tree.cpp"
+  "libgambatte/src/mem/snes_spc/dsp.cpp"
```
Perhaps we should consider making snes_spc a separate "library" in this regard instead of bundled into libgambatte (not really any different in practice since it'd always just a static library included in libgambatte). Although that's probably suited for a later time.

FtZPetruska committed:
Add vcpkg config for compatibility with VS.

FtZPetruska committed:
Update documentation to cover Visual Studio.

FtZPetruska committed:
Properly set symbols' visibility.

- Classes and functions in public headers should be exported.
- Internal symbols should hidden.
- The "cinterface.cpp" module was left unmodified for compatibility.
- This reduces binary size for non-Windows platforms.
- For Windows, it allows using a DLL with the public API.

FtZPetruska committed:
Add the libgambatte::libgambatte alias.

This allows using the same target when adding the project as a
subproject using `add_subdirectory`.

FtZPetruska committed:
Test building with pkg-config more thoroughly.

FtZPetruska committed:
Move misplaced GBEXPORT.

FtZPetruska commented:
Regarding symbols visibility, as it stood, the public API (`gambatte.h`) was not available when linking against the DLL.

On the other hand, the `cinterface.cpp` file explicitly exported functions despite them not be declared in any header.

To address that, I made the following changes in https://github.com/pokemon-speedrunning/gambatte-core/pull/19/commits/4ea626ad66ba8134aef373587ce0ffe5625663a1:
- Made a public `gbexport.h` header that defines a `GBEXPORT` macro similar to what was defined in `cinterface.h`, with the GCC/Clang equivalent.
- Exported every class/function declared in the public headers
- Switch the Linux/macOS behaviour to match what Windows does (symbols hidden by default), this has the side-effect of slightly reducing the binary size too.
- The `cinterface.cpp` file was left "as-is" for backwards compatibility. Since the `GBEXPORT` macro no longer contains the `extern "C"` part, it was added explicitly around these functions.

This change makes the shared library usable on Windows as previously programs such as the `testrunner` would fail to link dynamically.

CasualPokePlayer commented:
Is this GBEXPORT on classes behavior for if you're using a shared library yet you're using the C++ API? You shouldn't be using that API directly in such a case, as that usually will end up resulting in linking errors and trying to dynamically link with dlopen/LoadLibrary would not exactly be fun to do (both due to compiler specific name mangling). You should only be using the C++ API if you're using the library as a static library (and you are in fact a C++ user and using the same compiler here).

To properly address the issue of a lack of a header, function declarations should just be added to cinterface.h (or whatever it's named now) for the cinterface.cpp functions. Programs such as the testrunner shouldn't be dynamically linking anyways (those statically link unless something changed in this PR), so something is going wrong much earlier there.

FtZPetruska commented:
Just to make sure I get this correctly:
- We do **not** want the C++ API to be exported (`gambatte::GB/NewState*/PakInfo` classes), these constructs should only be usable from a static library.
- When using the shared library, only the functions defined in `cinterface.cpp` should be accessible.

> To properly address the issue of a lack of a header, function declarations should just be added to cinterface.h (or whatever it's named now) for the cinterface.cpp functions.

I am thinking about making it a `gambatte-c.h` header, and add the prototypes for the functions defines currently in `cinterface.cpp`. I think it may desirable to also build these functions in the static library, so they are always available regardless of the library type (in scons, `cinterface.cpp` was only included in shared builds).

> Programs such as the testrunner shouldn't be dynamically linking anyways

I had made the static library a requirement to run the test:
```cmake
if(NOT LIBGAMBATTE_BUILD_STATIC)
  message(FATAL_ERROR "Building the static library is required to run tests")
endif()
```

But it's good to know that you are not supposed to dynamically link it.

FtZPetruska committed:
Also include cinterface.cpp in static builds.

FtZPetruska committed:
Partially revert 4ea626ad66ba8134aef373587ce0ffe5625663a1

The C++ API should *not* be exported in the shared library.

FtZPetruska committed:
Expose the cinterface functions in gambatte-c.h

CasualPokePlayer commented on libgambatte/include/gambatte-c.h:
```diff
@@ -0,0 +1,160 @@
+#define GBEXPORT
+#endif
+
+extern "C" {
```
Since this might be expected to be #included from a C context (not C++), it may be better to #ifdef this around `__cplusplus`.

Similarly, instead of #including gambatte.h or newstate.h, it may be better to just forward declare `struct GB` instead, (and other redeclarations from the gambatte namespace as that might not be available in a C context), at the very least do that when `__cplusplus` is not defined.

CasualPokePlayer commented on libgambatte/include/gambatte-c.h:
```diff
@@ -0,0 +1,160 @@
+#include "newstate.h"
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+#define GBEXPORT __declspec(dllexport)
```
If this is used by the user of a dll rather than the dll itself, then dllexport would be incorrect, dllimport is the correct declspec (and I don't think visibility attribrute is correct either for non-Windows cases, rather it should just be `extern`

FtZPetruska replied:
Checking GCC's documentation on [visibility](https://gcc.gnu.org/wiki/Visibility), you're correct about the dllexport/dllimport.

According to that same documentation, for the non-Windows case, `__attribute__((visibility("default")))` seems to indicate both export and import.

FtZPetruska replied:
Seems like the change in https://github.com/pokemon-speedrunning/gambatte-core/pull/19/commits/d1ae9ec4a2bb31f66ae7b75f2287b607aa8a7004 [broke CI](https://github.com/FtZPetruska/gambatte-core/actions/runs/5167158457/jobs/9307863094). 

Looking around, other projects such as SDL seem to only be using [dllexport](https://github.com/libsdl-org/SDL/blob/SDL2/include/begin_code.h#L56).

CasualPokePlayer replied:
Breaking CI likely indicates it got used internally in the library, or it got used when linking it in statically. You can only use dllimport when you are a user of the library (so not the library itself) and the library is a dll (if it's static, you don't use dllimport/dllexport).

FtZPetruska committed:
Make gambatte-c usable in C.

FtZPetruska committed:
Use dllexport/import properly.

dllexport is only meant to be used when building the library. When using
the API, dllimport should be used.

CasualPokePlayer commented on libgambatte/include/gambatte-c.h:
```diff
@@ -16,145 +16,171 @@
+#include <stdbool.h>
+
+typedef struct GB GB;
+typedef struct FPtrs FPtrs;
```
FPtrs is expected to be constructed from the user side rather than the dll side, so it should be fully defined here.

FtZPetruska committed:
Move FPtrs to a C-compatible header.

FtZPetruska committed:
Only define dllexport/import with the shared lib.

CasualPokePlayer commented on libgambatte/include/fptrs.h:
```diff
@@ -0,0 +1,21 @@
+
+#ifdef __cplusplus
+namespace gambatte {
+extern "C" {
```
extern "C" here is a bit interesting, *technically* this is correct, although in practice it typically doesn't actually matter, as most (if not all practical) C++ implementations wouldn't have any issues with a C++ linkage function being assigned to a C linkage function pointer and vice versa (as they function the same at runtime), although this isn't actually guaranteed by the standard (C++ linkage could entail a different calling convention compared to the same function being assigned C linkage, in practice this never occurs).

On that, technically this is not correctly handled in gambatte.h as they all would use C++ (default) linkage, so if we want that to be ""correctly"" done, all those callback definitions need extern "C".

FtZPetruska replied:
On the off chance that there is a compiler out there that defaults to using a different calling convention between C and C++, I went ahead and declared the other function prototypes as `extern "C"` for consistency.

Moreover, to make sure the C API is actually usable from C, I added a simple test in [c75fe83](https://github.com/pokemon-speedrunning/gambatte-core/pull/19/commits/c75fe837cbcbe7b857fd74783525538a287e8cc0) that creates and deletes a GB object using the C API.

FtZPetruska committed:
Add a simple test for the C API.

This ensures the header is usable from a C source.

FtZPetruska committed:
Typedef FPtrs for when it's used from C.

FtZPetruska committed:
Define callback functions with C linkage.

This is mainly for correctness, but ensures compatibility when __stdcall
is the default.
